File: src/main/java/alfheim/common/integration/minetweaker/handler/MTHandlerManaInfuser.java
package alfheim.common.integration.minetweaker.handler;

import static alfheim.common.integration.minetweaker.MinetweakerAlfheimConfig.*;

import java.util.ArrayList;

import alfheim.api.AlfheimAPI;
import alfheim.api.ModInfo;
import alfheim.api.crafting.recipe.RecipeManaInfuser;
import minetweaker.IUndoableAction;
import minetweaker.MineTweakerAPI;
import minetweaker.api.item.IIngredient;
import minetweaker.api.item.IItemStack;
import net.minecraft.item.ItemStack;
import stanhebben.zenscript.annotations.ZenClass;
import stanhebben.zenscript.annotations.ZenMethod;

@ZenClass("mods." + ModInfo.MODID + ".ManaInfuser")
public class MTHandlerManaInfuser {

	@ZenMethod
    public static void addRecipe(IItemStack output, int mana, IIngredient... inputs) {
		MineTweakerAPI.apply(new Add(new RecipeManaInfuser(mana, getStack(output), getObjects(inputs))));
	}
	
	@ZenMethod
	public static void removeRecipe(IItemStack output) {
		MineTweakerAPI.apply(new Remove(getStack(output)));
	}
	
	private static class Add implements IUndoableAction {

		private final RecipeManaInfuser recipe;
		
		public Add(RecipeManaInfuser recipeManaInfuser) {
			recipe = recipeManaInfuser;
		}

		@Override
		public void apply() {
			AlfheimAPI.addInfuserRecipe(recipe);
		}

		@Override
		public boolean canUndo() {
			return true;
		}

		@Override
		public void undo() {
			AlfheimAPI.removeInfusionRecipe(recipe);
		}

		@Override
		public String describe() {
			return String.format("Adding Mana Infuser recipe %s", recipe);
		}

		@Override
		public String describeUndo() {
			return String.format("Removing Mana Infuser recipe %s", recipe);
		}

		@Override
		public Object getOverrideKey() {
			return null;
		}
	}
	
	private static class Remove implements IUndoableAction {

		private final ItemStack output;
		final ArrayList<RecipeManaInfuser> removed = new ArrayList<RecipeManaInfuser>();
		
		public Remove(ItemStack stack) {
			output = stack;
		}

		@Override
		public void apply() {
			RecipeManaInfuser rec;
			do {
				removed.add(rec = AlfheimAPI.removeInfusionRecipe(output));
			} while (rec != null);
		}

		@Override
		public boolean canUndo() {
			return true;
		}

		@Override
		public void undo() {
			for (RecipeManaInfuser rec : removed) AlfheimAPI.addInfuserRecipe(rec);
		}

		@Override
		public String describe() {
			return String.format("Removing all Mana Infuser recipes for %s", output.getUnlocalizedName());
		}

		@Override
		public String describeUndo() {
			return String.format("Re-adding previously removed Mana Infuser recipes for %s", output.getUnlocalizedName());
		}

		@Override
		public Object getOverrideKey() {
			return null;
		}
	}
}

File: src/main/java/alfheim/common/integration/minetweaker/handler/MTHandlerManaInfuser.kt
package alfheim.common.integration.minetweaker.handler

import alfheim.api.*
import alfheim.api.crafting.recipe.RecipeManaInfuser
import alfheim.common.integration.minetweaker.MinetweakerAlfheimConfig.getObjects
import alfheim.common.integration.minetweaker.MinetweakerAlfheimConfig.getStack
import minetweaker.*
import minetweaker.api.item.*
import net.minecraft.item.ItemStack
import stanhebben.zenscript.annotations.*
import java.util.*

@ZenClass("mods." + ModInfo.MODID + ".ManaInfuser")
object MTHandlerManaInfuser {
	
	@ZenMethod
	fun addRecipe(output: IItemStack, mana: Int, vararg inputs: IIngredient) {
		MineTweakerAPI.apply(Add(RecipeManaInfuser(mana, getStack(output), *getObjects(inputs))))
	}
	
	@ZenMethod
	fun removeRecipe(output: IItemStack) {
		MineTweakerAPI.apply(Remove(getStack(output)))
	}
	
	private class Add(private val recipe: RecipeManaInfuser): IUndoableAction {
		
		override fun apply() {
			AlfheimAPI.addInfuserRecipe(recipe)
		}
		
		override fun canUndo(): Boolean {
			return true
		}
		
		override fun undo() {
			AlfheimAPI.removeInfusionRecipe(recipe)
		}
		
		override fun describe(): String {
			return String.format("Adding Mana Infuser recipe %s", recipe)
		}
		
		override fun describeUndo(): String {
			return String.format("Removing Mana Infuser recipe %s", recipe)
		}
		
		override fun getOverrideKey(): Any? {
			return null
		}
	}
	
	private class Remove(private val output: ItemStack): IUndoableAction {
		internal val removed = ArrayList<RecipeManaInfuser>()
		
		override fun apply() {
			var rec: RecipeManaInfuser?
			do {
				rec = AlfheimAPI.removeInfusionRecipe(output)
				removed.add(rec!!)
			} while (rec != null)
		}
		
		override fun canUndo(): Boolean {
			return true
		}
		
		override fun undo() {
			for (rec in removed) AlfheimAPI.addInfuserRecipe(rec)
		}
		
		override fun describe(): String {
			return String.format("Removing all Mana Infuser recipes for %s", output.unlocalizedName)
		}
		
		override fun describeUndo(): String {
			return String.format("Re-adding previously removed Mana Infuser recipes for %s", output.unlocalizedName)
		}
		
		override fun getOverrideKey(): Any? {
			return null
		}
	}
}